File: clients/net-chat-go/internal/bitchat/protocol/packet.go
package protocol

import (
	"encoding/hex"
	"time"
)

// PacketFlags is the 1-byte flags word in the header. Bits mirror
// mainline packet.rs. Fields are exported so the sign/verify helpers
// in higher layers can clear has_signature before hashing.
type PacketFlags struct {
	HasRecipient bool
	HasSignature bool
	IsCompressed bool
}

// PacketFlagsFromByte decodes a flags byte.
func PacketFlagsFromByte(b byte) PacketFlags {
	return PacketFlags{
		HasRecipient: b&0x01 != 0,
		HasSignature: b&0x02 != 0,
		IsCompressed: b&0x04 != 0,
	}
}

// ToByte encodes flags for the wire.
func (f PacketFlags) ToByte() byte {
	var b byte
	if f.HasRecipient {
		b |= 0x01
	}
	if f.HasSignature {
		b |= 0x02
	}
	if f.IsCompressed {
		b |= 0x04
	}
	return b
}

// Packet is the on-wire packet framed by BinaryProtocol. Nil recipient
// / signature slices signal "not present"; the encoder consults the
// flag AND the slice — matches Rust's Option<[u8; N]> shape.
type Packet struct {
	Version     byte
	MessageType MessageType
	TTL         byte
	Timestamp   time.Time
	Flags       PacketFlags
	SenderID    [8]byte
	RecipientID *[8]byte // nil ⇒ not present
	Payload     []byte
	Signature   *[64]byte // nil ⇒ not present
}

const (
	// HeaderSize is the fixed header prefix: version + type + ttl +
	// timestamp(8) + flags + payload_len(2) = 13 bytes.
	HeaderSize = 13
	// MaxPayloadSize is the 2-byte payload length upper bound.
	MaxPayloadSize = 65535
)

// New builds a fresh outbound packet with defaults matching packet.rs::new.
func New(mt MessageType, sender [8]byte, payload []byte) *Packet {
	return &Packet{
		Version:     1,
		MessageType: mt,
		TTL:         3,
		Timestamp:   time.Now().UTC(),
		SenderID:    sender,
		Payload:     payload,
	}
}

// WithRecipient sets the recipient and lifts the has_recipient flag.
func (p *Packet) WithRecipient(rid [8]byte) *Packet {
	p.RecipientID = &rid
	p.Flags.HasRecipient = true
	return p
}

// WithSignature stamps a signature and lifts the has_signature flag.
func (p *Packet) WithSignature(sig [64]byte) *Packet {
	p.Signature = &sig
	p.Flags.HasSignature = true
	return p
}

// SenderIDHex returns the 16-char lowercase-hex form.
func (p *Packet) SenderIDHex() string { return hex.EncodeToString(p.SenderID[:]) }

// RecipientIDHex returns the 16-char lowercase-hex form of the
// recipient if present, empty string otherwise.
func (p *Packet) RecipientIDHex() string {
	if p.RecipientID == nil {
		return ""
	}
	return hex.EncodeToString(p.RecipientID[:])
}

File: clients/net-chat-android/app/src/main/java/io/grfx/netchat/protocol/Packet.java
package io.grfx.netchat.protocol;

/**
 * BitChat wire packet. Encoder/decoder in {@link BinaryProtocol}.
 *
 * <p>{@link #recipientId} and {@link #signature} are optional — null
 * signals "not present". The flag bits in {@link Flags} MUST agree
 * with the null-ness of these fields; {@link BinaryProtocol#encode}
 * checks this.
 */
public final class Packet {

    /** 1-byte flags word: has_recipient (bit 0), has_signature (bit 1), is_compressed (bit 2). */
    public static final class Flags {
        public boolean hasRecipient;
        public boolean hasSignature;
        public boolean isCompressed;

        public static Flags fromByte(byte b) {
            Flags f = new Flags();
            f.hasRecipient = (b & 0x01) != 0;
            f.hasSignature = (b & 0x02) != 0;
            f.isCompressed = (b & 0x04) != 0;
            return f;
        }

        public byte toByte() {
            int v = 0;
            if (hasRecipient) v |= 0x01;
            if (hasSignature) v |= 0x02;
            if (isCompressed) v |= 0x04;
            return (byte) v;
        }
    }

    public static final int HEADER_SIZE = 13;      // version + type + ttl + ts(8) + flags + len(2)
    public static final int MAX_PAYLOAD_SIZE = 65535;

    public byte version = 1;
    public MessageType messageType;
    public byte ttl = 3;
    public long timestampMillis;                   // ms since epoch (BE u64 on wire)
    public final Flags flags = new Flags();
    public byte[] senderId;                        // exactly 8 bytes
    public byte[] recipientId;                     // 8 bytes, or null
    public byte[] payload;                         // arbitrary length; ≤ MAX_PAYLOAD_SIZE encoded
    public byte[] signature;                       // 64 bytes, or null

    public Packet() {}

    /** Convenience builder — sets timestamp to now, ttl=3. */
    public static Packet outbound(MessageType type, byte[] senderId, byte[] payload) {
        Packet p = new Packet();
        p.messageType = type;
        p.senderId = senderId;
        p.payload = payload;
        p.timestampMillis = System.currentTimeMillis();
        return p;
    }

    /** Set recipient AND lift the has_recipient flag in one step. */
    public Packet withRecipient(byte[] rid) {
        this.recipientId = rid;
        this.flags.hasRecipient = true;
        return this;
    }

    /** Set signature AND lift the has_signature flag in one step. */
    public Packet withSignature(byte[] sig) {
        this.signature = sig;
        this.flags.hasSignature = true;
        return this;
    }
}
